### PR TITLE
Translations for the mods

### LINKED ISSUE #2219: Mod translation
Mod translation links returns 404 - Not found. Further instructions missing too. How this feature works? I placed .properties file with translation into "jsons/translations" and not working.

### PR DESCRIPTION
**Pull request by JackRainy**
Resolves #2219 

The changes are:
* Newly generated language files contain emply lines separation between key elements (leftover from #2205 )
* The language files in the mods folders are used for the translations
* The `TranslationFileWriter` is simplified a bit: no redundant calls - no caching needed
* Percentage calculation is done on-the-fly while forming the new files (instead of an extra run as it was before)
* Running the app with `rewriteTranslationFiles = true` will also create the translation files for the mods (if any found). The additional template files can be also used, if needed.

### CODE REVIEW

**Comment by Iver88** [comment]
Thank you @JackRainy I can't wait to implementing it to my mods :)

File: core/src/com/unciv/models/translations/TranslationFileWriter.kt [hunk 1]
```diff
@@ -17,99 +18,130 @@ import java.lang.reflect.Field
...
+            for (line in linesFromTemplates) {
+                if (!line.contains(" = ")) {
+                    // small hack to insert empty lines
+                    if (line.startsWith(specialNewLineCode))
+                        stringBuilder.appendln()
+                    else // copy as-is
+                        stringBuilder.appendln(line)
+                    continue
+                } else // count translatable lines only once (e.g. for English)
+                    if (language == "English") countOfTranslatableLines++
```
Comment (yairm210) [inline core/src/com/unciv/models/translations/TranslationFileWriter.kt]: I don't understand this - if the line doesn't contain " = ", meaning it's not a translation line, then we count it for English?

File: core/src/com/unciv/models/translations/TranslationFileWriter.kt [hunk 2]
```diff
@@ -17,99 +18,130 @@ import java.lang.reflect.Field
...
+        val countOfTranslatedLines = HashMap<String, Int>()
+        // iterate through all available languages
+        for (language in translations.getLanguages()) {
+            var translationsOfThisLanguage = 0
+            val stringBuilder = StringBuilder()
+
+            for (line in linesFromTemplates) {
+                if (!line.contains(" = ")) {
+                    // small hack to insert empty lines
+                    if (line.startsWith(specialNewLineCode))
```
Comment (yairm210) [inline core/src/com/unciv/models/translations/TranslationFileWriter.kt]: Why not just have it an empty line, as it has been so far?
It'll enter if(!line.contains(" = "), and be copied as-is - that is to say, empty

Comment (yairm210) [reply core/src/com/unciv/models/translations/TranslationFileWriter.kt]: Ahhh nevermind I see what you did below

File: core/src/com/unciv/models/translations/TranslationFileWriter.kt [hunk 3]
```diff
@@ -17,99 +18,130 @@ import java.lang.reflect.Field
...
-    private fun generateTutorialsStrings(): Collection<String> {
 
-        if (generatedStrings.containsKey("Tutorials"))
-            return generatedStrings.getValue("Tutorials")
 
-        generatedStrings["Tutorials"] = mutableSetOf()
-        val tutorialsStrings = generatedStrings["Tutorials"]
+    private fun generateTutorialsStrings(): MutableSet<String> {
+
+        val tutorialsStrings = mutableSetOf<String>()
```
Comment (yairm210) [inline core/src/com/unciv/models/translations/TranslationFileWriter.kt]: If you want them to be in the samee order you put them in, this should be a LinkedHashSet

File: core/src/com/unciv/models/translations/TranslationFileWriter.kt [hunk 4]
```diff
@@ -17,99 +18,130 @@ import java.lang.reflect.Field
...
-        if (generatedStrings.isNotEmpty())
-            return // do not regenerate if the strings are ready
+        // Using LinkedHashMap (instead of HashMap) is important to maintain the order of sections in the translation file
+        val generatedStrings = LinkedHashMap<String, MutableSet<String>>()
 
+        var uniqueIndexOfNewLine = 0
         val jsonParser = JsonParser()
-        val folderHandler = Gdx.files.internal("jsons")
+        val folderHandler = if (modFolder != null) modFolder.child("jsons")
+                             else Gdx.files.internal("jsons")
```
Comment (yairm210) [inline core/src/com/unciv/models/translations/TranslationFileWriter.kt]: I notice this is the third time for this pattern - may I suggest fun getLocation(subPath:String, modFolder:String?) = if(modfolder=null) modFolder.child(subPath) else Gdx.files.internal(subPath)

File: core/src/com/unciv/models/translations/TranslationFileWriter.kt (hunk 1 above)
Comment (JackRainy) [reply core/src/com/unciv/models/translations/TranslationFileWriter.kt]: Vise versa. It is located in the `else` section of the `! line.contains(" = ")` condition.
So if the line _contains_ " = ", it is a translation line, then we count it for English.

File: core/src/com/unciv/models/translations/TranslationFileWriter.kt (hunk 3 above)
Comment (JackRainy) [reply core/src/com/unciv/models/translations/TranslationFileWriter.kt]: That's true, and it is, because `mutableSetOf` is declared as:
> public inline fun <T> mutableSetOf(): MutableSet<T> = LinkedHashSet()

File: core/src/com/unciv/models/translations/TranslationFileWriter.kt (hunk 1 above)
Comment (JackRainy) [reply core/src/com/unciv/models/translations/TranslationFileWriter.kt]: I have inverted the if-condition in order to make it more understandable.

File: core/src/com/unciv/models/translations/TranslationFileWriter.kt (hunk 4 above)
Comment (JackRainy) [reply core/src/com/unciv/models/translations/TranslationFileWriter.kt]: Good suggestion, thank you.

File: core/src/com/unciv/models/translations/TranslationFileWriter.kt [hunk 5]
```diff
@@ -82,8 +87,7 @@ object TranslationFileWriter {
 
             countOfTranslatedLines[language] = translationsOfThisLanguage
 
-            val fileWriter = if (modFolder != null) modFolder.child(languageFileLocation.format(language))
-                                else Gdx.files.local(languageFileLocation.format(language))
+            val fileWriter = getFileHandle(modFolder,languageFileLocation.format(language))
```
Comment (yairm210) [inline core/src/com/unciv/models/translations/TranslationFileWriter.kt]: After looking over this, it looks like I made a mistake.
It's as follows - on Android, there are 2 file locations - internal and local, where internal is the assets packed into the apk and local is files added afterwards, not from the apk.
Internal is read-only, and local is read-write.
In Desktop, where you don't have that separation, they are both in the same location.

HOWEVER, even if you're in Desktop, *by definition* the internal location is read-only, so if you want to edit the assets you need to use the _local_ file handler.

tl;dr local required for writing on Desktop, internal required for reading on Android, which means that this specific part (which does the writing) needs to access the files.local....

My bad

Comment (yairm210) [reply core/src/com/unciv/models/translations/TranslationFileWriter.kt]: I'm curious, though - does it work when you try running it? If it does, then maybe the restriction for "internal is read only" is only active on Android, and the writing code (which is active only on Desktop) can stay as-is

Comment (JackRainy) [reply core/src/com/unciv/models/translations/TranslationFileWriter.kt]: I have noticed that sometimes it is `internal` and sometimes it is `local` but did not pay much attention to it since any of these worked well on Desktop.
As far as I understand you, it is enough to replace read-only `internal` for read-write `local`.

Comment (yairm210) [reply core/src/com/unciv/models/translations/TranslationFileWriter.kt]: That should be enough, since we're in the TranslationWriter class
In the TranslationReader we'll need to use `internal`

**Comment by Iver88** [comment]
Are city names from mod translatable as well?

**Comment by Iver88** [comment]
I tested it and not working. What is wrong?

Snímek obrazovky 2020-03-25 v 0 50 48

**Comment by rh-github-2015** [comment]
> I tested it and not working. What is wrong?
> 
        for(modFolder in Gdx.files.local("mods").list()) {
            val modTranslationFile = modFolder.child(translationFileName)

means the files should not go into a 'translations' subfolder.

**Comment by rh-github-2015** [comment]
I saved this earlier in the wrong place:

As I read the reader code,
a) mod translations would need to go directly into the mod root not a /translations/ subfolder
b) All mods present would be active, whether or not the mod is included in the current game
c) mods be read in OS api order, which usually is undefined. Therefore, should conflicts exist (e.g. a mod developer keeps different versions of the same mod in parallel for whatever reason), it might actually surprise someone which file actually wins.
d) It might be desirable to allow mods to 'overrule' vanilla translations - this should at least be predictable. I didn't check, but did you consider?

**Comment by yairm210** [comment]
a. Correct, good catch, we'll need to fix that
b. Also correct, but that's fine by me - it has to do with 'when do we load translations', and currently we load translations when the game is loading.
We could change this to add the translations for mods only when starting the game, but I don't hink it's that important now
c. HMMMM Indeed
d. Well, that's basically how the current mod translations work... =)

**Comment by yairm210** [comment]
Wait, no, I reread and a. is actually not correct. We write to, and read from, the translations subfolder.
But we'll need to check if/why it isn't working.

**Comment by yairm210** [comment]
AHHHH I think I found A problem, not necessarily THE problem:

image
image

Changing from local to internal, and of course the mod isn't in internal storage (on Android at leat)

**Comment by Iver88** [comment]
A. So "translation" subfolder or root? I came out from standard game source structure.
C. Like in Farming Simulator or Surviving Mars games, you can name mod folder as aSomething, aaSomething, zSomething, zzSomething etc. at least. I think this can be addressed after/if will be implemented simple mod manager with possibility to change mods order. Mainly on Android is this mod manager needed.

**Comment by yairm210** [comment]
A. Translation subfolder. Mod structure should be identical to main game structure.

**Comment by Iver88** [comment]
Ok. I have prepared my mod since yesterday. I'll test it in free time. Or you can get it here for testing - https://github.com/Iver88/Unciv-Lands-of-the-Bohemian-Crown

**Comment by yairm210** [comment]
image
Looks like mod-specific translations are working

image

And so is translation file generation, if you enable rewriteTranslationFiles

**Comment by Iver88** [comment]
What about possibility to load mod name from mod config with translation support so it will show translated in mod list?

Snímek obrazovky 2020-03-25 v 11 39 27

**Comment by Iver88** [comment]
@yairm210 rewriteTranslationFiles? Where?

**Comment by Iver88** [comment]
It uses default name of civ unique instead of translated version from mod "uniqueName".

Snímek obrazovky 2020-03-25 v 11 45 57

**Comment by rh-github-2015** [comment]
> It uses default name of civ unique instead of translated version from mod "uniqueName".

uniqueName is a corpse. dead. Was changed sometime I guess so only the key referencing enum class UniqueAbility does anything.

You'd have to translate the description given there but then all nations using that unique key would get the new description.

Something I also noticed days ago - no way to mod-add nation uniques.

But that correlates with my issue (?)... wait a moment...

**Comment by Iver88** [comment]
More info tile string is not translated.

Snímek obrazovky 2020-03-25 v 11 54 58

**Comment by rh-github-2015** [comment]
> More info tile string is not translated.

You mean the city name? Should be simple to find/fix. Gimme an hour or several.

**Comment by Iver88** [comment]
Doesn't takes global translation of civ uniques when not translated in mod.

Snímek obrazovky 2020-03-25 v 12 05 50

**Comment by yairm210** [comment]
This is good, we're adding general-purpose polish as well with these changes =)

**Comment by rh-github-2015** [comment]
@yairm210 Did you fix the _name_ of fun TileInfo.toString() while you did e0f14a5? When I stumbled on it I thought - now waitaminute - I'd always expect this when overloaded to yield a _technical_ identification of the object for debugging and logs... I'm not really serious here.

@Iver88  > Doesn't takes global translation of civ uniques when not translated in mod.
I guess I would need more precise info. Are the mod files for that screen shot exactly as committed to https://github.com/Iver88/Unciv-Lands-of-the-Bohemian-Crown? Where? Civilopedia? For which entity - a unit I assume?

**Comment by Iver88** [comment]
@rh-github-2015 When selecting nation and in civilopedia as well. It should takes translation at least from original translation when not defined in mod. In Czech.properties in (official translation) it is translated already. These uniques are from original game.

**Comment by yairm210** [comment]
There are 2 problems here

1. You shoved all of the uniques into 1 string, when they should be different strings:
image

2. The translation generation doesn't filter out terms that already exist in the 'main' translations, meaning that even if everything was fine, you'd have 'double' lines - just like you have with the Knight and Chariot lines.

**Comment by Iver88** [comment]
1. Nevermind, I am dumb. It's working now. Common coding misteake :)

**Comment by rh-github-2015** [comment]
I was puzzled needlessly as well. The screenshot is from civilopedia.nation: the 'lost ability vs the unit it replaces' description exists only on the nation description (Nation.kt addUniqueUnitsText), not in the entry for the unit itself (BaseUnit getDescription).

Might be worth to consolidate this. There's no reason not to display such a defining property on the unit page as well.

@Iver88 : Nation: introduction, afterPeace, tradeRequest are Inca leftovers :)

**Comment by JackRainy** [comment]
Wow! I have missed quite a lot of fun here.
Sorry, can't follow the replies whole the day.
Is there anything to be fixed after all, or I missed the party?

**Comment by JackRainy** [comment]
Do we still consider having the nation uniques as an enum class good idea?
It seems the previous solution having them in Nations.json was more flexible one.

**Comment by Iver88** [comment]
> @Iver88 : Nation: introduction, afterPeace, tradeRequest are Inca leftovers :)

Ok. Anything else? It has some purpose for Inca mod or it can be deleted from Inca mod as well?

**Comment by Iver88** [comment]
> Wow! I have missed quite a lot of fun here.
> Sorry, can't follow the replies whole the day.
> Is there anything to be fixed after all, or I missed the party?

Only missing stuff is support for translation of mod name. But it is icing on the cake :) Guys did great work on code for modding support.

**Comment by Iver88** [comment]
@JackRainy And possibility to translate uniqueName in mod's Nations.json. But it is harder as @rh-github-2015 said.

**Comment by JackRainy** [comment]
> And possibility to translate uniqueName in mod's Nations.json.

It was possible before uniqueNames were moved out of Nations,json and got replaced with enum. So the fix might be in rolling that change back :)

**Comment by rh-github-2015** [comment]
But I still think having multiple effects triggered by one key as a large string is inferior to having a nation be a container for a collection of simpler effects. Would ease sharing of code with social policies. As I sad, then nation/civ/policies would be structurally identical to baseunit/mapunit/promotions and could be covered by common code........
Nation unique *description* as one big novel and translatable as a whole would then be one way, listing the effects another for civilopedia display. In the former case, bringing back uniqueDescription but not using it as a key would be following logically.

To take a step back, de-coupling a mod label from the folder name would be good: one extra translatable property per nation. Relatively straightforward. Who volunteers? I have a twrp session to nurse.
Next: Leave the enum but stop using its descriptions, move them back to json properties. Automatically makes mods able to abuse an existing "unique" under a different disguise. Maybe replace the strings in the enum with "unused" or remove them entirely, but leave the keys. Also simple coding...
The "effect" system I keep blathering about can be treated independently if at all.

**Comment by rh-github-2015** [comment]
> > @Iver88 : Nation: introduction, afterPeace, tradeRequest are Inca leftovers :)
> Ok. Anything else? It has some purpose for Inca mod or it can be deleted from Inca mod as well?

These are _your_ lines and not very fitting for Charles the something-th. No relation to Incas other than being overlooked copies. That's all.

**Comment by JackRainy** [comment]
> Only missing stuff is support for translation of mod name.

Done within PR #2260

**Comment by JackRainy** [comment]
> de-coupling a mod label from the folder name would be good: one extra translatable property per nation. Relatively straightforward. 

I agree, it makes sense to de-couple. However, disagree that it should be part of Nations. There can be a mod without new nations, e.g. UltimateSciencePack or something like that.

**Comment by Iver88** [comment]
> > > @Iver88 : Nation: introduction, afterPeace, tradeRequest are Inca leftovers :)
> > > Ok. Anything else? It has some purpose for Inca mod or it can be deleted from Inca mod as well?
> 
> These are _your_ lines and not very fitting for Charles the something-th. No relation to Incas other than being overlooked copies. That's all.

It's not edited in my mod right now. So I'll get rid of it :)

**Comment by rh-github-2015** [comment]
> disagree that it should be part of Nations.
Facepalm - of course. I *was* thinking that one extra json with just one property might be overkill before commenting that, but then dementia hit. Your solution is definitely preferrable.

**Comment by Iver88** [comment]
City is not translated when renaming it.

Snímek obrazovky 2020-04-02 v 1 12 22

**Comment by JackRainy** [comment]
That's expected, because you can name it as something not translatable.
Or did you rename it to something which already has a translation in your translation file?

**Comment by Iver88** [comment]
It's original name Vratislav is Wroclaw in czech.
